### DISCUSSION TITLE
Override conversion for STL-like user types

### DISCUSSION THREAD

**Discussion by Numendacil** [nlohmann/json#3699] [Q&A] [answered]
Suppose i have the following struct `foo`, which defines a `begin` and `end` function for the convenience of ranged-for loops
```cpp
struct foo
{
    std::string tag;
    std::vector<int> data;
    ...
    iterator begin() { return this->data.begin(); }
    iterator end() { return this->data.end(); }
    // so i can use for (const auto& p : foo), etc
};
```
Now if i try to define a custom conversion to json by
```cpp
void from_json(const json& j, foo& p)
{
    p.tag = j.at("tag").get<std::string>();
    p.data = j.at("data").get<std::vector<int>>();
    ...
}
```
It would not work, since the existence of iterator functions would made the library use the default serialization as if `std::vector<int>` is present, i.e. `get<foo>()` is like `get<std::vector<int>>()` with the results stored in `foo::data`.
Therefore my question is that is there a way to override this default behaviour or i have to choose between ranged-for and custom conversions?

**Comment by falbrechtskirchinger** [comment, marked as answer]
A related issue is very high on my to-do list. I'll take your use case into account when I get to it.

You can probably work around the issue by implementing a custom serializer that forwards to the default serializer for all types but `foo`.
